### ISSUE TITLE
[BUG] Node affinity remains on the PV after setting data locality to disabled

### ISSUE THREAD

**Issue by jani888** [longhorn/longhorn#9966]
## Describe the bug

After changing the data locality of a PV from `strinct-local` to `disabled` or `best-effort` the node affinity selector and the `VolumeAttributes:      dataLocality=strict-local` configs remain on the PV. This results in pods unable to start.

## To Reproduce

1. Create a volume with `strict-local` data locality.
2. Change the data locality to `disabled`
3. Notice that the node affinity remains on the PV

## Expected behavior

After changing the node affinity from `strict-local` to `disabled`, the node affinity should be removed.

## Support bundle for troubleshooting

Due to compliance reasons I'm not allowed to share the support bundle.

## Environment

 - Longhorn version: v1.7.2
 - Installation method: HELM
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 5
 - Node config
   - OS type and version: Ubuntu
   - Kernel version: -
   - CPU per node: 8
   - Memory per node: 32G
   - Disk type (e.g. SSD/NVMe/HDD): NVMe
   - Network bandwidth between the nodes (Gbps): 1
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Proxmox
 - Number of Longhorn volumes in the cluster: 5

**Comment by derekbit** [comment]
> Change the data locality to disabled

How do you change the data locality?

**Comment by jani888** [comment]
I stopped the workload, detached the volume using the UI, and on the UI changed the data locality

**Comment by derekbit** [comment]
The `VolumeAttributes` in a PV are derived from the parameters in the associated StorageClass and are immutable. When you modify the value of `volume.spec.dataLocality`, the updated value cannot be reflected in the `VolumeAttributes` (this is the limitation of PV). However, the Longhorn system still respects the updated value and behaves accordingly.

**Comment by jani888** [comment]
Yes, but this way, the node affinity is not removed from the PV, and the scheduler won't schedule my pod to a different node, and I cannot move the volume or the workload. Is there any solution for this? I think this worked in older versions.

**Comment by derekbit** [comment]
> Yes, but this way, the node affinity is not removed from the PV, and the scheduler won't schedule my pod to a different node, and I cannot move the volume or the workload. Is there any solution for this? I think this worked in older versions.

Got it. It is due to https://github.com/longhorn/longhorn-manager/pull/1933.

One workaround you can try is 
1. **IMPORTANT: Back up the volume for ensuring any surprise**
2. Detach the volume
3. Dump the PV data by `kubectl get pv <PV_NAME> -o yaml > pv.yaml` and remove the
   ```
	spec:
	  nodeAffinity: {...}
   ``` 
   from the yaml file.
4. Delete the PV by `kubectl delete pv <PV_NAME>`. If it is stuck, edit the PV and remove the finalizers.
5. Recreate the PV by `kubectl apply -f pv.yaml`

**Comment by jani888** [comment]
Thank you so much, this is what I was looking for. 

Will the fix be in the next release?

**Comment by c3y1huang** [comment]
I am not sure if there is anything that Longhorn could do here, as Kubernetes restricts updating the node affinity once it is set.
https://github.com/kubernetes/kubernetes/blob/v1.32.2/pkg/apis/core/validation/validation.go#L2132-L2135
And I am not sure about the potential side effects if Longhorn forcefully deletes the PV.

**Comment by derekbit** [comment]
> I am not sure if there is anything that Longhorn could do here, as Kubernetes restricts updating the node affinity once it is set. https://github.com/kubernetes/kubernetes/blob/v1.32.2/pkg/apis/core/validation/validation.go#L2132-L2135 And I am concerned about the potential side effects if Longhorn forcefully deletes the PV.
> 
> /cc [@derekbit](https://github.com/derekbit)

As @c3y1huang's investigation, let's close the ticket.

**Comment by jani888** [comment]
I appreciate the investigation, but I don't think closing this issue as "wontfix" is a good resolution. If this limitation cannot be addressed due to Kubernetes restrictions, then the UI should reflect this to prevent misleading users.

Warning in the UI: When setting data locality to strict-local, there should be a clear warning that switching back to disabled or best-effort will not remove node affinity, potentially preventing pod scheduling.
UI Restriction: The option to switch from strict-local to disabled or best-effort should be disabled in the UI, or at least come with a clear warning about the consequences.
Right now, the UI allows changing the setting without indicating that node affinity will persist, leading to unexpected behavior and requiring manual PV modifications as a workaround. This is not an intuitive user experience.

If Longhorn cannot fix the underlying issue, then improving the UI to prevent users from running into this problem would be a reasonable alternative. Can this be reconsidered?

**Comment by jani888** [comment]
> And I am concerned about the potential side effects if Longhorn forcefully deletes the PV.

I understand the concern about the potential side effects of forcefully deleting the PV, but this was the exact workaround that was previously suggested as a solution. If manually deleting and recreating the PV is considered safe enough to recommend, then Longhorn should ideally automate this process (with proper safeguards) rather than leaving it as a manual workaround.

**Comment by derekbit** [comment]
> I appreciate the investigation, but I don't think closing this issue as "wontfix" is a good resolution. If this limitation cannot be addressed due to Kubernetes restrictions, then the UI should reflect this to prevent misleading users.

@jani888 Sounds good to me. However, the warning message cannot be delivered to users if they configure data locality via a CR update. Since this is a limitation of Kubernetes, we do not intend to work around it in Longhorn. Before finding a better way solution, what we can do 

- Have a warning message on UI (optional)
- Have a warning or notice on official document 
- Create a knowledge base (KB) doc for the workaround

WDYT @jani888 @c3y1huang @innobead
